device.go

```diff
@@ -0,0 +1,144 @@
+//
+// Copyright (c) 2018 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/kata-containers/agent/pkg/uevent"
+	pb "github.com/kata-containers/agent/protocols/grpc"
+	"github.com/sirupsen/logrus"
+)
+
+type deviceDriversHandler func(device pb.Storage, spec *pb.Spec) error
+
+var deviceDriversHandlerList = map[string]deviceDriversHandler{
+	"blk": blockDeviceHandler,
+}
+
+func blockDeviceHandler(device pb.Storage, spec *pb.Spec) error {
+	if spec.Linux == nil || len(spec.Linux.Devices) == 0 {
+		return fmt.Errorf("No devices found from the spec, cannot update")
+	}
+
+	// First need to make sure the expected device shows up properly,
+	// and then we need to retrieve its device info (such as major and
+	// minor numbers), useful to update the device provided
+	// through the OCI specification.
+	if err := waitForDevice(device.Source); err != nil {
+		return err
+	}
+
+	stat := syscall.Stat_t{}
+	if err := syscall.Stat(device.Source, &stat); err != nil {
+		return err
+	}
+
+	major := int64(stat.Rdev / 256)
+	minor := int64(stat.Rdev % 256)
+
+	agentLog.WithFields(logrus.Fields{
+		"device-path":  device.Source,
+		"device-major": major,
+		"device-minor": minor,
+	}).Info("handling block device")
+
+	// Update the spec
+	updated := false
+	for idx, d := range spec.Linux.Devices {
+		if d.Path == device.MountPoint {
```

Comment:
`d.Path` in the OCI spec is the device path. How do you expect it to be a `MountPoint` path here?

It seems that you really need `d.Path == device.Source` instead? But again, this function only works for block devices that are also present in the storage spec (which is also useful since agent can mount the block devices for containers), whereas the commit message and #132 both intend to provide a generic way to pass host devices to the sandbox.

Reply:
No no, I need `MountPoint` because this is the expected destination for the device, and we expect this destination to be the source of the device which has been passed-through.

Reply:
In OCI spec, `spec.Linux.Devices.Path` is a device path like `/dev/sdb` and libcontainer would call `mknodat` to create it with the provided device numbers. While in kata agent protocol, `pb.Storage.MountPoint` the is mount point of the storage which is a directory outside of the container rootfs. How do you expect them to be equal?

Reply:
@sboeuf I guess we have some misunderstanding here. Can you please provide an example config of `spec.Linux.Devices` and `pb.Storage` that you intend to handle with this PR?

Reply:
Ok so let me try to clarify here. In case we want to pass-through a block device such as `/dev/sdb` from the host to the container, we need to hotplug this device into the VM. After it has been hotplugged, it will show up as `/dev/vdX` name that we can anticipate from the runtime. That's why I expect the runtime to provide the right name `/dev/vdX` through the `Source` field, and the name expected by the OCI spec through the `Mountpoint` field, so that we can match with the spec which entry we should modify. Once we have found the device `/dev/vdX`, we get the __real__ major/minor numbers and we update them into the OCI spec for the corresponding device. If we don't update them, libcontainer will create a new device inside the container based on no real device, which means the device will be there but not linked to the real device.

Reply:
In that case, `Mountpoint` is not really a mount point of the device, but a device name inside the container. If we do want to make it so, I'd suggest changing the name from `Mountpoint` to something like `Target`, and document that it can either be a mount point of the device, or a device name to be represented inside the container, depending on the value of `Storage.Driver`.

And there must be 1:1 mapping between `pb.Storage.Target` and `spec.Linux.Devices`. Maybe we should make `pb.Storage.Target` a slice so that a device and be represented at multiple places in a container or to multiple containers in the sandbox?